File: class30/java/BD3001-ConexaoOracle.java
/*
 * BD3001-ConexaoOracle.java
 * Aula 30 - Banco de Dados - UA12 (Conectividade)
 *
 * Demonstra a conexao com Oracle Database via JDBC (driver "thin", Tipo 4),
 * execucao de uma consulta parametrizada e fechamento correto dos recursos
 * usando try-with-resources.
 *
 * Tabela de apoio usada nos exemplos (criar previamente no schema de teste):
 *
 *   CREATE TABLE funcionarios (
 *       id       NUMBER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
 *       nome     VARCHAR2(100) NOT NULL,
 *       cargo    VARCHAR2(50)  NOT NULL,
 *       salario  NUMBER(10,2)
 *   );
 *
 *   INSERT INTO funcionarios (nome, cargo, salario) VALUES ('Ana Souza', 'Veterinario', 6500.00);
 *   INSERT INTO funcionarios (nome, cargo, salario) VALUES ('Bruno Lima', 'Recepcionista', 2800.00);
 *   COMMIT;
 *
 * Dependencia Maven:
 *   <dependency>
 *       <groupId>com.oracle.database.jdbc</groupId>
 *       <artifactId>ojdbc11</artifactId>
 *       <version>23.4.0.24.05</version>
 *   </dependency>
 *
 * Compilar/rodar (fora de um projeto Maven, com o .jar do driver no classpath):
 *   javac -cp ojdbc11.jar BD3001-ConexaoOracle.java
 *   java  -cp .:ojdbc11.jar BD3001ConexaoOracle
 */

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class BD3001ConexaoOracle {

    // Ajuste host/porta/service name conforme o ambiente de laboratorio.
    private static final String URL = "jdbc:oracle:thin:@localhost:1521/XEPDB1";
    private static final String USUARIO = "app_user";
    private static final String SENHA = "senha_segura";

    public static void main(String[] args) {
        // Desde JDBC 4.0 (Java 6+) o registro do driver via Class.forName() e opcional:
        // o DriverManager descobre "oracle.jdbc.OracleDriver" automaticamente via
        // Service Provider Interface (META-INF/services/java.sql.Driver dentro do jar).
        // A linha abaixo e mantida comentada apenas por clareza/compatibilidade legada:
        // Class.forName("oracle.jdbc.OracleDriver");

        String cargoBuscado = "Veterinario";
        buscarFuncionariosPorCargo(cargoBuscado);
    }

    private static void buscarFuncionariosPorCargo(String cargo) {
        String sql = "SELECT id, nome, cargo, salario FROM funcionarios WHERE cargo = ?";

        // try-with-resources: Connection, PreparedStatement e ResultSet sao fechados
        // automaticamente ao final do bloco, mesmo se ocorrer excecao.
        try (Connection conexao = DriverManager.getConnection(URL, USUARIO, SENHA);
             PreparedStatement comando = conexao.prepareStatement(sql)) {

            comando.setString(1, cargo);

            try (ResultSet resultado = comando.executeQuery()) {
                System.out.println("Funcionarios com cargo = " + cargo + ":");
                System.out.println("----------------------------------------");

                int total = 0;
                while (resultado.next()) {
                    int id = resultado.getInt("id");
                    String nome = resultado.getString("nome");
                    String cargoBd = resultado.getString("cargo");
                    double salario = resultado.getDouble("salario");

                    System.out.printf("%d | %s | %s | R$ %.2f%n", id, nome, cargoBd, salario);
                    total++;
                }

                if (total == 0) {
                    System.out.println("Nenhum funcionario encontrado para o cargo informado.");
                }
            }

        } catch (SQLException e) {
            // SQLException e sempre "checked" em JDBC - erros de conexao, SQL invalido,
            // violacao de constraint etc. chegam por aqui.
            System.err.println("Erro ao conectar/consultar o Oracle Database.");
            System.err.println("SQLState: " + e.getSQLState() + " | Codigo: " + e.getErrorCode());
            System.err.println("Mensagem: " + e.getMessage());
        }
    }
}

File: class30/python/BD3001-ConexaoOracle.py
"""
BD3001-ConexaoOracle.py
Aula 30 - Banco de Dados - UA12 (Conectividade)

Demonstra a conexao com Oracle Database via python-oracledb (sucessor do
cx_Oracle), em modo "thin" (sem exigir Oracle Instant Client instalado),
usando gerenciadores de contexto (with) para garantir o fechamento da
conexao e do cursor, alem de um exemplo de connection pool.

Tabela de apoio (criar previamente no schema de teste):

    CREATE TABLE funcionarios (
        id       NUMBER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
        nome     VARCHAR2(100) NOT NULL,
        cargo    VARCHAR2(50)  NOT NULL,
        salario  NUMBER(10,2)
    );

Instalacao:
    pip install oracledb
"""

import oracledb

# Ajuste usuario/senha/dsn conforme o ambiente de laboratorio.
USUARIO = "app_user"
SENHA = "senha_segura"
DSN = "localhost:1521/XEPDB1"  # host:porta/service_name


def buscar_funcionarios_por_cargo(cargo):
    """Conecta, executa uma consulta parametrizada e imprime o resultado.

    Usa `with` tanto na conexao quanto no cursor: ao sair do bloco `with`,
    ambos sao fechados automaticamente, mesmo se ocorrer uma excecao.
    """
    sql = "SELECT id, nome, cargo, salario FROM funcionarios WHERE cargo = :cargo"

    try:
        with oracledb.connect(user=USUARIO, password=SENHA, dsn=DSN) as connection:
            with connection.cursor() as cursor:
                cursor.execute(sql, cargo=cargo)

                print(f"Funcionarios com cargo = {cargo}:")
                print("-" * 40)

                total = 0
                for id_, nome, cargo_bd, salario in cursor:
                    print(f"{id_} | {nome} | {cargo_bd} | R$ {salario:.2f}")
                    total += 1

                if total == 0:
                    print("Nenhum funcionario encontrado para o cargo informado.")

    except oracledb.DatabaseError as erro:
        (erro_obj,) = erro.args
        print("Erro ao conectar/consultar o Oracle Database.")
        print(f"Codigo: {erro_obj.code} | Mensagem: {erro_obj.message}")


def exemplo_com_pool():
    """Exemplo de uso de connection pool nativo do python-oracledb.

    Util quando a aplicacao faz varias operacoes ao longo do tempo (ex.: uma
    API web) - evita abrir/fechar uma conexao fisica a cada requisicao.
    """
    pool = oracledb.create_pool(
        user=USUARIO, password=SENHA, dsn=DSN, min=2, max=10, increment=1
    )

    try:
        with pool.acquire() as connection:
            with connection.cursor() as cursor:
                cursor.execute("SELECT COUNT(*) FROM funcionarios")
                (total,) = cursor.fetchone()
                print(f"Total de funcionarios cadastrados: {total}")
    finally:
        pool.close()


if __name__ == "__main__":
    buscar_funcionarios_por_cargo("Veterinario")
    print()
    exemplo_com_pool()
